shelley/chain-and-ledger/executable-spec/test/Rules/TestUtxo.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module Rules.TestUtxo where
+
+import           Data.Word (Word64)
+
+import           Hedgehog (Property, forAll, property, withTests)
+
+import           Control.State.Transition.Generator (ofLengthAtLeast, trace)
+import           Control.State.Transition.Trace (pattern SourceSignalTarget, signal, source,
+                     sourceSignalTargets, target)
+
+import           Coin (pattern Coin)
+import           TxData (pattern Tx, _wdrls)
+import           UTxO (balance)
+
+import           LedgerState (pattern UTxOState)
+import           MockTypes (UTXO)
+
+import           Test.Utils (assertAll)
+
+------------------------------
+-- Constants for Properties --
+------------------------------
+
+numberOfTests :: Word64
+numberOfTests = 300
+
+traceLen :: Word64
+traceLen = 100
+
+--------------------------
+-- Properties for UTXOW --
+--------------------------
+
+-- | Property that checks that the fees are non-decreasing
+feesNonDecreasing :: Property
+feesNonDecreasing = withTests (fromIntegral numberOfTests) . property $ do
+  tr <- fmap sourceSignalTargets $ forAll (trace @UTXO traceLen `ofLengthAtLeast` 1)
+
+  assertAll feesDoNotIncrease tr
+
+  where feesDoNotIncrease (SourceSignalTarget
+                            { source = UTxOState _ _ fees _
+                            , target = UTxOState _ _ fees' _}) =
+          fees <= fees'
+
+-- | Property that checks that the sum of the pots circulation, deposits and
+-- fees increases by the sum of withdrawals of a transaction.
+potsSumIncreaseWdrls :: Property
+potsSumIncreaseWdrls = withTests (fromIntegral numberOfTests) . property $ do
+  tr <-
```

Comment:
We might be consistent in the way we format this. We have two styles in this module, and I have used a third one:

```haskell
 tr <- fmap sourceSignalTargets
      $ forAll
      $ trace @DELEG traceLen `ofLengthAtLeast` 1
```

Did I win the "ant lover of the month" price yet?

Reply:
one version is used when we need to access the environment from the trace